### ISSUE TITLE
New Lint: temporary lock guards in if let / while let / match scrutinees

### ISSUE THREAD

**Issue by spidermana** [rust-lang/rust-clippy#16889]
### What it does

Recently, I am coding on some rust projects and noticed an easy mistake: a temporary lock guard can be held longer than expected when it is created inside the scrutinee of `if let`, `while let`, or `match`. This pattern is easy to run into when working with collections using methods like pop().

I also did a quick pattern search on GitHub and found several similar cases in real code, so this seems to be a practical footgun.

I’d be happy to work on a PR if this sounds like a good fit for Clippy.

## What it does

Warns when a temporary lock guard created in the scrutinee of `if let`, `while let`, or `match`
causes the lock to be held for longer than it appears.

This is a common pitfall: the lock is acquired in an expression that visually looks small and local, yet the actual drop scope spans the entire construct. The 2024 Edition partially mitigates this by [shortening the drop scope of temporary values](https://doc.rust-lang.org/edition-guide/rust-2024/temporary-if-let-scope.html#footnote-scrutinee), but the issue can still arise in the if body.

#### `if let`

```rust
if let Some(item) = list.lock().unwrap().pop() {
    process_item(item);
}
```

This could be written as:

```rust
let item = {
    let mut guard = list.lock().unwrap();
    guard.pop()
};

if let Some(item) = item {
    process_item(item);
}
```

#### `while let`

```rust
while let Some(item) = list.lock().unwrap().pop() {
    process_item(item);
}
```

This could be written as:

```rust
while let Some(item) = {
    let mut guard = list.lock().unwrap();
    guard.pop()
} {
    process_item(item);
}
```

#### `match`

```rust
match list.lock().unwrap().pop() {
    Some(item) => process_item(item),
    None => {}
}
```

This could be written as:

```rust
let item = {
    let mut guard = list.lock().unwrap();
    guard.pop()
};

match item {
    Some(item) => process_item(item),
    None => {}
}
```

It also applies to collections like:
- Vec with pop()
- VecDeque with pop_front() / pop_back()
- BinaryHeap with pop()
- BTreeSet with pop_first() / pop_last()
- ...

So in all of these cases, the returned value is owned, so the guard usually does not need to remain alive that long.

By contrast, if the method returns a reference, like `front()`, then the guard must stay alive, so that is a different case and should usually not be linted.

## Why is this bad?

Temporary values created inside larger expressions are dropped at the end of the statement.
For lock guards, this can mean that the mutex or rwlock remains held during the body of an
`if let`, `while let`, or `match`, which is easy to overlook and can lead to performance problems, unnecessary lock contention, or deadlocks in worse cases.

### Category

nursery or suspicious

### Suggested implementation

- Target `if let`, `while let`, and `match`
- Only lint known lock guard types first:
  - `std::sync::MutexGuard`
  - `std::sync::RwLockReadGuard`
  - `std::sync::RwLockWriteGuard`

If people think this is reasonable, I’d be glad to take it on and open a PR. Feedback on the implementation, category, and naming is very welcome.

### Advantage

Makes lock scope explicit, improve the performance and helps prevent a subtle class of lock-holding bugs.

### Drawbacks

May have some false positives, not sure yet.
But this seems like a good fit for `nursery` or `suspicious`  first.

#### `if let`

```rust
if let Some(item) = list.lock().unwrap().pop() {
    process_item(item);
}
```

This could be written as:

```rust
let item = {
    let mut guard = list.lock().unwrap();
    guard.pop()
};

if let Some(item) = item {
    process_item(item);
}
```

#### `while let`

```rust
while let Some(item) = list.lock().unwrap().pop() {
    process_item(item);
}
```

This could be written as:

```rust
loop {
    let item = {
        let mut guard = list.lock().unwrap();
        guard.pop()
    };

    let Some(item) = item else {
        break;
    };

    process_item(item);
}
```

#### `match`

```rust
match list.lock().unwrap().pop() {
    Some(item) => process_item(item),
    None => {}
}
```

This could be written as:

```rust
let item = {
    let mut guard = list.lock().unwrap();
    guard.pop()
};

match item {
    Some(item) => process_item(item),
    None => {}
}
```

### Comparison with existing lints

This seems related to #11773 and #5219 (`clippy::if_let_mutex`), but those cases do not cover `while let` or `match`. Also, my proposed lint is especially relevant for collections-processing code such as `pop()`-based examples above.

### Additional Context

To some extent, this kind of issue can easily arise when using collections, since they make the code more simple. It is natural to take the first value returned by pop and use it directly in the scrutinee expression.

**Comment by QEEK-AI** [comment]
**QEEK.AI Agent:** 

"Thank you for this well-thought-out proposal! This is indeed a practical footgun that can cause subtle performance issues and deadlocks. The examples with `pop()` on collections are particularly compelling since the guard visually appears to be short-lived."

### Suggested Implementation Approach

**1. Lint Category**
`suspicious` seems appropriate as the initial category, with potential to graduate to `correctness` or `performance` after dogfooding. The 2024 Edition mitigations do reduce the severity for some cases, but as you noted, the body of `if let` remains problematic.

**2. Detection Strategy**

Consider expanding beyond just the three lock guard types to include ecosystem types:
- `parking_lot::{MutexGuard, RwLockReadGuard, RwLockWriteGuard, ...}`
- `std::sync::{MutexGuard, RwLockReadGuard, RwLockWriteGuard}`
- Potentially `tokio::sync` guards for async code

You can use `clippy_utils::ty::is_type_diagnostic_item` or `match_def_path` for detection.

**3. AST Pattern to Target**

```rust
// Target these expressions:
if let Some(x) = expr.lock().unwrap().method() { ... }
while let Some(x) = expr.lock().unwrap().method() { ... }
match expr.lock().unwrap().method() { ... }
```

**4. Key Implementation Consideration: Owned vs. Borrowed**

Your distinction between owned methods (like `pop()`) vs. reference-returning methods (like `front()`) is crucial. Consider:

```rust
// Should lint (returns owned T):
vec.lock().unwrap().pop()
vec.lock().unwrap().remove(0)
heap.lock().unwrap().pop()

// Should NOT lint (returns reference, guard must stay alive):
vec.lock().unwrap().front()
vec.lock().unwrap().get(0)
vec.lock().unwrap().last()
```

You may need to check the return type's lifetime relationship to the receiver to determine if the guard must outlive the scrutinee.

**5. Alternative Suggestion Format**

Your rewrite suggestions are clear. Consider also suggesting a block-based approach:

```rust
// Suggested fix style:
if let Some(item) = list.lock().unwrap().pop() { ... }

// Suggestion:
let item = list.lock().unwrap().pop();
if let Some(item) = item { ... }
```

This is simpler than the block-scoped guard version when the user doesn't need the guard for anything else.

### Edge Cases to Consider

1. **Reference returns**: As noted, methods returning `&T` require the guard to stay alive
2. **Guard used in pattern**: `if let Some(Guard(ref x)) = ...` cases
3. **Drop order dependencies**: Code that relies on the extended drop scope
4. **2024 Edition**: Ensure the lint accounts for the new temporary scope rules

### Comparison with Existing Lints

You're right that this extends beyond `if_let_mutex`. That lint focuses on the specific pattern of locking inside an `if let` condition without considering the temporary lifetime issue. Your proposal is more general and catches the `while let` and `match` variants.

### Next Steps

This looks like a great addition! Some suggestions before opening a PR:

1. **Name suggestion**: Consider `temporary_lock_guard_in_scrutinee` or `extended_lock_guard_lifetime`
2. Run `cargo dev new_lint` to scaffold the implementation
3. Check for existing utility functions in `clippy_utils` for detecting lock guard types
4. Include comprehensive test cases covering the edge cases above

**Would you like early feedback on a draft PR? Happy to help review the implementation approach.**
